rust/bridge/node/futures/tests-node-module/src/store_like.rs

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright 2020 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use futures::try_join;
+use neon::prelude::*;
+use std::panic::AssertUnwindSafe;
+use std::sync::Arc;
+
+use signal_neon_futures::*;
+
+struct NameStore {
+    js_queue: EventQueue,
+    store_object: Arc<Root<JsObject>>,
+}
+
+impl NameStore {
+    fn new<'a>(cx: &mut FunctionContext<'a>, store: Handle<'a, JsObject>) -> Self {
+        Self {
+            js_queue: cx.queue(),
+            store_object: Arc::new(store.root(cx)),
+        }
+    }
+
+    async fn get_name(&self) -> Result<String, String> {
+        let store_object_shared = self.store_object.clone();
+        JsFuture::get_promise(&self.js_queue, move |cx| {
+            let store_object = store_object_shared.to_inner(cx);
+            let op = store_object
+                .get(cx, "getName")?
+                .downcast_or_throw::<JsFunction, _>(cx)?;
+            let result = op
+                .call(cx, store_object, std::iter::empty::<Handle<JsValue>>())?
+                .downcast_or_throw(cx)?;
+            store_object_shared.finalize(cx);
+            Ok(result)
+        })
+        .then(|cx, result| match result {
+            Ok(value) => match value.downcast::<JsString, _>(cx) {
+                Ok(s) => Ok(s.value(cx)),
+                Err(_) => Err("name must be a string".into()),
+            },
+            Err(error) => Err(error
+                .to_string(cx)
+                .expect("can convert to string")
+                .value(cx)),
+        })
+        .await
+    }
+}
+
+impl Finalize for NameStore {
+    fn finalize<'a, C: Context<'a>>(self, cx: &mut C) {
+        self.store_object.finalize(cx)
+    }
+}
+
+async fn double_name_from_store_impl(store: &mut NameStore) -> Result<String, String> {
+    Ok(format!(
+        "{0} {1}",
+        store.get_name().await?,
+        store.get_name().await?
+    ))
+}
+
+// function doubleNameFromStore(store: { getName: () => Promise<string> }): Promise<string>
+pub fn double_name_from_store(mut cx: FunctionContext) -> JsResult<JsObject> {
+    let js_store = cx.argument(0)?;
+    let mut store = NameStore::new(&mut cx, js_store);
+
+    promise(&mut cx, async move {
+        let future = AssertUnwindSafe(double_name_from_store_impl(&mut store));
```

Comment:
@jack-signal This AssertUnwindSafe is the last thing I'm worried about for our real uses: all our store APIs take `&mut store`, which means that Rust assumes they're *not* unwind-safe futures. Yet we're catching panics at FFI boundaries. I could drop the requirement for UnwindSafe on the `promise` method (I put it there myself) but it kind of makes sense as a reminder? Not sure.

Reply:
No strong opinion here but I don't see a problem with having to annotate with `AssertUnwindSafe` as a reminder to think about what else you are doing during the call.